### ISSUE TITLE
No data warning stays if data lazy loaded and gridOptions changed

### ISSUE THREAD

**Issue by zewa666** [ghiscoding/Angular-Slickgrid#1340]
### Describe the bug

There is an issue that if you don't configure enableEmptyDataWarningMessage in your gridOptions, the message is properly shown if data load is lazy (since true is the default), but the message won't go away after data arrives and gridOptions got modified meanwhile.

This bug happens because the check to toggle the visibility off is made here https://github.com/ghiscoding/Angular-Slickgrid/blob/312be48230a3b847809fac90dd46bf1b1af9828e/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L703 and expects the default enableEmptyDataWarningMessage to be applied.

Now if the user modifies the gridOptions again meanwhile, those defaults will be missed and the second time the check won't work. An example where I'm doing that is within the postProcess of OData in order to update totals and other criteria that might change due to freshly loaded permission sets.

### Reproduction

* make sure enableEmptyDataWarningMessage is not defined in your gridOptions
* delay the initial data load
* no data message properly appears
* once the data loads, message still stays around

Easiest way to reproduce is to apply the following patch to Example01
```patch
--- a/src/app/examples/grid-basic.component.ts
+++ b/src/app/examples/grid-basic.component.ts
@@ -53,10 +53,14 @@ export class GridBasicComponent implements OnInit {
     };

     // mock some data (different in each dataset)
-    this.dataset1 = this.mockData(NB_ITEMS);
+    setTimeout(() => this.dataset1 = this.mockData(NB_ITEMS), 1000);
     this.dataset2 = this.mockData(NB_ITEMS);
   }

+  ngAfterViewInit() {
+    this.gridOptions1 = { ...this.gridOptions1, gridWidth: 900 };
+  }
+
```

### Expectation

TL;DR
Instead of overwriting the REF here, we should update it.
E.g.
```typescript
Object.assign(this.gridOptions, this.mergeGridOptions(this.gridOptions));
```
https://github.com/ghiscoding/Angular-Slickgrid/blob/312be48230a3b847809fac90dd46bf1b1af9828e/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L482

Long story:
I think what I'd really expect is that my gridOptions passed in to the component
```
[gridOptions]="gridOptions"
```

should be updated by ref, hence allowing me to continue modifying the local object without any issues.
I've tried alternatively to do

```
this.gridOptions = {...this.angularGrid.gridService._gridOptions, ...this.gridOptions};
```
but that results in an TS issue due to difference of the viewComponent property within rowDetailView. Fixable via an any cast, but ugly as hell. Plus it's super error prone to have to think to always obtain the options via the gridService instead your local ref.

### Environment Info

```shell
latest version of angular-slickgrid
```

### Validations

- [X] Follow our [Code of Conduct](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/.github/CODE_OF_CONDUCT.md)
- [X] Read the [HOWTO - Quick Start](https://ghiscoding.gitbook.io/angular-slickgrid/getting-started/quick-start).
- [X] Check that there isn't [already an issue](https://github.com/ghiscoding/Angular-Slickgrid/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/ghiscoding/Angular-Slickgrid/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by ghiscoding** [comment]
> Instead of overwriting the REF here, we should update it.

It had other indirect consequences which I forgot about, but yes I do a deep copy of the options internally

**Comment by zewa666** [comment]
thats true but the property becomes overwritten inside the angularslickgrid component and thus doesnt propagate changes to the caller, e.g my component hosting it.

**Comment by ghiscoding** [comment]
Sure I know but I just don't remember why I chose to do the deep copy, I'm sure that there is a non-negligible side effect for both approach. I also know that some Cypress tests will fail, the best would be to try it out and see what happens. However this might be a breaking change for some and I just came out with a major release so I don't know, let's try and see.

**Comment by zewa666** [comment]
with regards to Cypress tests ... good news is no new tests are failing for me. Bad news is that even without the modifcation the current master breaks on:

* Example 26
  * should open Export->Excel context sub-menu then open Feedback->ContactUs sub-menus and expect previous Export menu to no longer exists
* Example 29
  * should have pop songs folder with updated aggregations including 4 pop songs of Sum(400.3MB) / Avg(66.72MB)
* Example 32
  *  should mouse over Task 2 cell and expect async tooltip to show
  *  should mouse over Task 6 cell and expect async tooltip to show

not sure if I messed up something else but perhaps you could give all the 590 tests a run and see what happens on your instance.

---

although the current approach feels ... wrong, you're right its a change of behavior. So I'd go with

```ts
this.gridOptions = this.angularGrid.gridService._gridOptions as GridOption;
```
in the onAngularGridCreated handler.

Perhaps this could be something to think about for a potential next major release

**Comment by ghiscoding** [comment]
We could add a new grid option that could be use internally to keep grid options deep copy or use simple extend when that option is enabled. 6pac (Ben) had an option called `mixinDefaults` that was doing pretty much what you're asking for, perhaps we can use that but again I think there was some failures with that option enabled because it's used for both GridOption and Columns, in SlickGrid he has this piece of code in SlickGrid constructor
```ts
if (options?.mixinDefaults) {
      // use provided options and then assign defaults
      if (!this._options) { this._options = options as O; }
      Utils.applyDefaults(this._options, this._defaults);
} else {
      this._options = Utils.extend<O>(true, {}, this._defaults, options);
}
```

For the Cypress tests, I had to use `testIsolation: false` since Cypress 10 made isolation the default, I thought that option was local to the each example though, I'm surprised that is global since I use `cy.visit()` but who knows maybe that's the reason.

**Comment by zewa666** [comment]
tbh naah ... the additional property is just overcomplicating things and still would be false by default. So you'd have to find it first than understand what it does, for those limited use-cases that you might roll into just to figure out that you could have done so yourself as well.

If anything the only alternative option to fixing the root-cause for me seems to be to document it better that the application of defaults aint gonna be shipped back to your provided obj.

**Comment by ghiscoding** [comment]
The new grid option was more meant like a temp solution until the next major, I was want to provide a temp solution before changing the behavior in the next major (that might take a while though since I'm pretty much done with everything)

I'm open to suggestion as long as it's not too much breaking change (behavior change).

**Comment by zewa666** [comment]
I'm with you on that. so as mentioned i'd cheer for improving the docs instead. I just dont know where it would be placed best. Quickstart looks a bit too prominent for such a detail

**Comment by ghiscoding** [comment]
but just going with docs still won't fix the problem. You could add a section under Grid Functionalities which is probably a good section for these kind things

**Comment by zewa666** [comment]
i just fear there is no proper way to address this without the breaking change `Object.assign(this.gridOptions, this.mergeGridOptions(this.gridOptions));`

since its not worth it and the user workaround is that simple I'd push the proper fix to later once another major is ready to ship.

the mixin approach, or really anothing other related to new properties wont cut it

**Comment by ghiscoding** [comment]
@zewa666 should this be closed by the PR #1344 or you want to leave this open because of the change to do in future major change?

**Comment by zewa666** [comment]
I'll close it and reopen later once the major change is around
